### ISSUE TITLE
PID and (succ PID) do not generate reliably unique temp directories.

### ISSUE THREAD

**Issue by creswick** [feuerbach/temporary#8]
This (`findTempName`) has a tendency to create temp directory names that conflict; particularly when some or all of the following conditions are true:

 - Multiple temp directories are created per process.
 - Multiple copies of the program are running at the same time (e.g., on a busy CI machine)
 - Multiple *haskell* programs using this library are running at the same time.

```Haskell
-- | Create a temporary directory.
createTempDirectory
  :: FilePath -- ^ Parent directory to create the directory in
  -> String -- ^ Directory name template
  -> IO FilePath
createTempDirectory dir template = do
  pid <- c_getpid
  findTempName pid
  where
    findTempName x = do
      let dirpath = dir </> template ++ show x
      r <- MC.try $ mkPrivateDir dirpath
      case r of
        Right _ -> return dirpath
        Left  e | isAlreadyExistsError e -> findTempName (x+1)
                | otherwise              -> ioError e
```

Here's an alternate version that should more reliably generate unique temp directories (I will try and create an MR with a tested version, but I can't do that at the moment):

```haskell
-- | Create a temporary directory.
createTempDirectory
  :: FilePath -- ^ Parent directory to create the directory in
  -> String -- ^ Directory name template
  -> IO FilePath
createTempDirectory dir template = do
  pid <- c_getpid
  findTempName pid 0
  where
    -- Separate the PID from the incremental tries when the directory name is not available:
    findTempName x tries = do
      let dirpath = dir </> template ++ show x ++ "_" ++ show tries
      r <- MC.try $ mkPrivateDir dirpath
      case r of
        Right _ -> return dirpath
        Left  e | isAlreadyExistsError e -> findTempName x (tries+1)
                | otherwise              -> ioError e
```

**Comment by Fuuzetsu** [comment]
That's very hardly any better isn't it? You're just increasing a different part of the name but still in the same fashion as all thr ither instances.

Why not just use a random number for the suffix? It will make conflicts much more rare.

**Comment by creswick** [comment]
You're right that a random suffix could reduce conflict, but I think simply separating the incremental suffix from the PID gets most of the way there -- and that's what I had time to implement ;)

I think it's worth more explicitly describing the problem with the current implementation, though, since that may not have been clear in my initial bug report.  My apologies if I'm beating a dead horse here:

Using a PID is great, in my opinion, since it does set a baseline for uniqueness in most situations, and it also makes it easy to link the directories on disk to specific processes.  That's handy for cleaning things up when processes crash (you can tell which directories are in use by comparing with PS, with reasonably high accuracy, for example).

However, for that to work, you *cannot* change the PID to resolve conflicts, because `PID + 1` is *also* a PID, and it's *not* the pid of the running process -- as soon as that's done, all of the benefits of using the PID are gone.

The additional field I added (`tries`) creates a hierarchy under the PID -- each process now uses a different counter to differentiate temp directories that that process created, and the only opportunity for conflict is if one process dies prematurely, and another process is started with the same PID (That also needs to be a Haskell process, using the same parent dir, and the same template -- for most purposes this means the same application has to be started up, using the same pid it had when it crashed previously.).

Even if that happens, the downside to the app is that it will just increment `tries` until it finds a free directory -- an operation linear in the number of temp directories that the program ever creates in one run.

The existing approach can cause a program to try as many new directories as there are processes running. We've had CI jobs block for a very, very long time because a long sequence of jobs was running, creating temp directories with incremented PIDs while an earlier process tried in vain to create a second or third temp directory, only to find that a dozens or hundreds of other processes had already created directories with those PIDs.

**Comment by UnkindPartition** [comment]
Thanks guys — I've also run into similar issues at work recently. I have some code lying around that generates random names, will merge soon.

**Comment by Fuuzetsu** [comment]
@creswick Right. I'm saying you have the same problem if you try to create lots of directories in same process. But I suppose that's more obvious issue to pin down. I still think random name is best and seems that's what's going to happen anyway.

**Comment by UnkindPartition** [comment]
This is now fixed in master. Could you check if it works for you?

**Comment by Fuuzetsu** [comment]
btw iirc I saw some code depending on the old naming scheme but it was awful and deserves breaking anyway

I would simply _remove_ the comment on how to the name is generated and give NO promises by design.

**Comment by UnkindPartition** [comment]
I don't consider it a promise (perhaps I should make it clear), but it may be useful to know when debugging etc.

**Comment by creswick** [comment]
Looks good to me, thanks!

**Comment by UnkindPartition** [comment]
Released as temporary-1.3 now.
